File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/controller/PurchasesController.java
package com.minimarket.api.controller;

import com.minimarket.api.dto.ApiMessageResponse;
import com.minimarket.api.dto.CreatePurchaseDto;
import com.minimarket.api.dto.PurchaseDto;
import com.minimarket.api.service.PurchaseService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/api/purchases")
public class PurchasesController {

    private final PurchaseService purchaseService;

    public PurchasesController(PurchaseService purchaseService) {
        this.purchaseService = purchaseService;
    }

    @GetMapping
    public List<PurchaseDto> getAll() {
        return purchaseService.getAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<PurchaseDto> getById(@PathVariable Integer id) {
        var purchase = purchaseService.getById(id);
        return purchase == null ? ResponseEntity.notFound().build() : ResponseEntity.ok(purchase);
    }

    @PostMapping
    public ResponseEntity<?> create(@RequestBody CreatePurchaseDto dto) {
        var result = purchaseService.create(dto);
        if (!result.success()) {
            return ResponseEntity.badRequest().body(new ApiMessageResponse(result.error()));
        }

        URI location = ServletUriComponentsBuilder.fromCurrentRequest()
            .path("/{id}")
            .buildAndExpand(result.data().id())
            .toUri();

        return ResponseEntity.created(location).body(result.data());
    }
}

File: backend/Minimarket.Api/Controllers/PurchasesController.cs
using Microsoft.AspNetCore.Mvc;
using Minimarket.Api.DTOs;
using Minimarket.Api.Services;

namespace Minimarket.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class PurchasesController(IPurchaseService purchaseService) : ControllerBase
{
    [HttpGet]
    public async Task<ActionResult<IReadOnlyCollection<PurchaseDto>>> GetAll()
        => Ok(await purchaseService.GetAllAsync());

    [HttpGet("{id:int}")]
    public async Task<ActionResult<PurchaseDto>> GetById(int id)
    {
        var purchase = await purchaseService.GetByIdAsync(id);
        return purchase is null ? NotFound() : Ok(purchase);
    }

    [HttpPost]
    public async Task<ActionResult<PurchaseDto>> Create([FromBody] CreatePurchaseDto dto)
    {
        var result = await purchaseService.CreateAsync(dto);
        if (!result.Success)
        {
            return BadRequest(new { message = result.Error });
        }

        return CreatedAtAction(nameof(GetById), new { id = result.Purchase!.Id }, result.Purchase);
    }
}
